core/arch/arm/plat-imx/pm/imx7_suspend.c

```diff
@@ -0,0 +1,183 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <arm.h>
+#include <arm32.h>
+#include <console.h>
+#include <drivers/imx_uart.h>
+#include <io.h>
+#include <imx.h>
+#include <imx_pm.h>
+#include <kernel/panic.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/generic_boot.h>
+#include <kernel/misc.h>
+#include <mm/core_mmu.h>
+#include <mm/core_memprot.h>
+#include <sm/sm.h>
+#include <sm/psci.h>
+#include <sm/pm.h>
+#include <stdint.h>
+
+/* i.MX7D has 2 cores, suspend/resume and low power idle share this */
+struct suspend_save_regs reg_stack[2];
+static int suspended_init;
+
+int imx7_cpu_suspend(uint32_t power_state __unused, uintptr_t entry,
+		     uint32_t context_id __unused, struct sm_nsec_ctx *nsec)
+{
+	uint32_t suspend_ocram_base = core_mmu_get_va(TRUSTZONE_OCRAM_START +
+						      SUSPEND_OCRAM_OFFSET,
+						      MEM_AREA_TEE_COHERENT);
+	struct imx7_pm_info *p = (struct imx7_pm_info *)suspend_ocram_base;
+	struct suspend_save_regs *regs;
+	int ret;
+
+	if (!suspended_init) {
+		imx7_suspend_init();
+		suspended_init = 1;
+	}
+
+	regs = &reg_stack[0];
+
+	/*
+	 * Store sec ctx regs
+	 * Now we are in SVC sec mode
+	 */
+	asm volatile (
+		"mov	r2, %0\r\n"
+		"mrs	r0, cpsr\r\n"
+		"mov	r3, #0x12\r\n" /* IRQ */
+		"mov	r4, #((1 << 6) | (1 << 7))\r\n" /* I | F */
+		"orr	r3, r3, r4\r\n"
+		"msr	cpsr, r3\r\n"
+		"mrs	r1, spsr\r\n"
+		"str	r1, [r2], #4\r\n"
+		"str	sp, [r2], #4\r\n"
+		"str	lr, [r2], #4\r\n"
+		"mov	r3, #0x11\r\n" /* FIQ */
+		"orr	r3, r3, r4\r\n"
+		"msr	cpsr, r3\r\n"
+		"mrs	r1, spsr\r\n"
+		"str	r1, [r2], #4\r\n"
+		"str	sp, [r2], #4\r\n"
+		"str	lr, [r2], #4\r\n"
+		"mov	r3, #0x1b\r\n"
+		"orr	r3, r3, r4\r\n"
+		"msr	cpsr, r3\r\n"
+		"mrs	r1, spsr\r\n"
+		"str	r1, [r2], #4\r\n"
+		"str	sp, [r2], #4\r\n"
+		"str	lr, [r2], #4\r\n"
+		"mov	r3, #0x17\r\n" /* ABT */
+		"orr	r3, r3, r4\r\n"
+		"msr	cpsr, r3\r\n"
+		"mrs	r1, spsr\r\n"
+		"str	r1, [r2], #4\r\n"
+		"str	sp, [r2], #4\r\n"
+		"str	lr, [r2], #4\r\n"
+		"mov	r3, #0x16\r\n" /* MON */
+		"orr	r3, r3, r4\r\n"
+		"msr	cpsr, r3\r\n"
+		"mrs	r1, spsr\r\n"
+		"str	r1, [r2], #4\r\n"
+		"str	sp, [r2], #4\r\n"
+		"str	lr, [r2], #4\r\n"
+		"msr	cpsr, r0\r\n"	/* SVC mode */
+		:
+		: "r" (regs)
+		: "r0", "r1", "r2", "r3", "r4"
+		);
+
+	ret = cpu_suspend((uint32_t)p,
+			  (int (*)(uint32_t))(suspend_ocram_base + sizeof(*p)));
+	/*
+	 * Sometimes cpu_suspend may not really suspended, we need to check
+	 * it's return value to restore reg or not
+	 */
+	if (ret < 0) {
+		DMSG("=== Not suspended, GPC IRQ Pending ===\n");
+		return 0;
+	}
+
+	plat_cpu_reset_late();
+
+	/* Restore register of different mode in secure world */
+	asm volatile (
```

Comment:
If `sm_save_modes_regs()` is used above, use `sm_restore_modes_regs(&nsec->mode_regs)` instead.

Reply:
This is restore secure world context. Ditto, I'll check using sm_restore_modes_regs.